src/Language/DifferentialDatalog/Attribute.hs

```diff
@@ -84,11 +88,20 @@ typedefValidateAttr d TypeDef{..} attr = do
                 $ "'sharef_ref' attribute is only applicable to extern types."
             check d (length tdefArgs == 1) (pos attr)
                 $ "Types annotated with 'shared_ref' must have exactly one type argument, e.g., \"Ref<'T>\"."
+         "dyn_alloc" -> do
+            check d (isNothing tdefType) (pos attr)
```

Comment:
Maybe you should have a helper function isExternType. Would make this more readable and abstract.